### PR TITLE
add RESULTS for kaldi pybind LF-MMI pipeline with PyTorch.

### PR DESCRIPTION
**Pull request by csukuangfj**
the training logs for kaldi pybind with PyTorch and nnet3 training are also contained.

kaldi pybind shares the same network architecture and `feats.scp` with nnet3. 

There are only two differences between kaldi pybind and nnet3:
(1) kaldi pybind uses BatchNorm to replace the first LDA layer
(2) kaldi pybind uses the optimizer from PyTorch.

WER/CER from kaldi nnet3 is better than kaldi pybind but kaldi pybind training with PyTorch is much faster.

The training time in total for 6 epochs is summarized as follows:
   - kaldi pybind with PyTorch: about 45 minutes
   - kaldi nnet3: about 4 hours 37 minutes == 277 minutes

It is possible that kaldi nnet3 can use less number of epochs to converge to a point that
has better CER/WER than kaldi pybind.

A very simple scheduler is used in PyTorch; the results for kaldi pybind may be improved
by using a better learning rate scheduler.

----

So what do we gain from kaldi pybind ?

1. training time. It is much faster
2. free to use various kinds of networks supported by PyTorch or it is very easy
  to write your own `nn.Module`.
3. you can try distributed training supported by Pytorch, e.g., DDP, or use horovod.
4. other fancy stuff limited by your imagination.

### CODE REVIEW

**Comment by csukuangfj** [comment]
@danpovey @naxingyu @jtrmal  @songmeixu @qindazhu 

please have a review.

**Comment by csukuangfj** [comment]
a screenshot of the tensorboard for kaldi pybind with PyTorch is shown as follows.

You can get this image after executing `run.sh`  and wait for the training stage to finish
and then run

```sh
tensorboard  --logdir ./exp/chain/train/tensboard
```

Screen Shot 2020-01-09 at 5 35 35 PM

**Comment by danpovey** [comment]
Also, can you please implement the delta+delta-delta feature extraction as part of the network?  This should improve the results.  You can follow recent Kaldi scripts for guidance.
And does the nnet3 baseline have i-vector?

**Comment by csukuangfj** [comment]
The nnet3 baseline uses NO ivector. It shares the same `feats.scp` with kaldi pybind.

**Comment by csukuangfj** [comment]
I will implement `delta+delta-delta` later.

**Comment by jtrmal** [comment]
closing and reopening to trigger the travis checks

**Comment by francisr** [comment]
What makes the Pytorch training so much faster than Kaldi's?

**Comment by csukuangfj** [comment]
@francisr 

The baseline network consists of
- convolution,
- batchnorm,
- full connected

I am not sure why PyTorch is such faster than kaldi. I guess GEMM/GEMV in PyTorch is highly optimized than kaldi.

**Comment by csukuangfj** [comment]
I will force push to remove the contained log files and you can find them here.

[kaldi-pybind-with-pytorch-training-log.txt](https://github.com/kaldi-asr/kaldi/files/4040241/kaldi-pybind-with-pytorch-training-log.txt)
[nnet3-training-log.txt](https://github.com/kaldi-asr/kaldi/files/4040243/nnet3-training-log.txt)

**Comment by RuABraun** [comment]
Surprised by how large the difference is! Awesome stuff.

**Comment by naxingyu** [comment]
> The nnet3 baseline uses NO ivector. It shares the same `feats.scp` with kaldi pybind.

Wait, aishell nnet3 baseline use ivector. Which baseline are you referring to?

**Comment by csukuangfj** [comment]
I was referring to this file
https://github.com/mobvoi/kaldi/blob/44ae951ea9c6f509dda24c60d29e5dddb482e3e1/egs/aishell/s10/local/run_tdnn_1b.sh#L100,

which has no ivector.

**Comment by csukuangfj** [comment]
closing and reopening to trigger travis CI.

**Comment by fanlu** [comment]
Modification of milestones in MultiStepLR optimizer from [2,6,8,9] to [1,2,3,4,5] will give 0.5~0.6 more precision.
```
==> exp/chain/decode_res_train_modelmodel_opadam_bs128_ep6_lr1e-3_fpe150_110_90_hn625_fpr1500000_ms1_2_3_4_5/test/scoring_kaldi/best
_cer <==                                                                                                                            
%WER 9.37 [ 9817 / 104765, 606 ins, 670 del, 8541 sub ] exp/chain/decode_res_train_modelmodel_opadam_bs128_ep6_lr1e-3_fpe150_110_90_
hn625_fpr1500000_ms1_2_3_4_5/test/cer_10_1.0                                                                                        
                                                                                                                                    
==> exp/chain/decode_res_train_modelmodel_opadam_bs128_ep6_lr1e-3_fpe150_110_90_hn625_fpr1500000_ms1_2_3_4_5/test/scoring_kaldi/best
_wer <==                                                                                                                            
%WER 18.16 [ 11701 / 64428, 1009 ins, 1926 del, 8766 sub ] exp/chain/decode_res_train_modelmodel_opadam_bs128_ep6_lr1e-3_fpe150_110_
90_hn625_fpr1500000_ms1_2_3_4_5/test/wer_12_0.5                                                                                     
==> exp/chain/decode_res_train_modelmodel_opadam_bs128_ep6_lr1e-3_fpe150_110_90_hn625_fpr1500000_ms1_2_3_4_5/dev/scoring_kaldi/best_
cer <==                                                                                                                             
%WER 7.69 [ 15790 / 205341, 668 ins, 801 del, 14321 sub ] exp/chain/decode_res_train_modelmodel_opadam_bs128_ep6_lr1e-3_fpe150_110_9
0_hn625_fpr1500000_ms1_2_3_4_5/dev/cer_9_1.0                                                                                        
                                                                                                                                    
==> exp/chain/decode_res_train_modelmodel_opadam_bs128_ep6_lr1e-3_fpe150_110_90_hn625_fpr1500000_ms1_2_3_4_5/dev/scoring_kaldi/best_
wer <==                                                                                                                             
%WER 15.89 [ 20293 / 127698, 2055 ins, 2733 del, 15505 sub ] exp/chain/decode_res_train_modelmodel_opadam_bs128_ep6_lr1e-3_fpe150_11
0_90_hn625_fpr1500000_ms1_2_3_4_5/dev/wer_10_0.0

```

**Comment by csukuangfj** [comment]
@fanlu 

Great to see that you can reproduce the results. 

I will update the parameters as you suggested.

**Comment by csukuangfj** [comment]
By the way, is your training time per epoch or per batch matches the log posted above?

**Comment by fanlu** [comment]
> By the way, is your training time per epoch or per batch matches the log posted above?

traning on P40 may consume more time than yours, it's about 66 minutes.
```
2020-01-14 14:12:00,333 INFO [train3.py:113] ./chain/train3.py --checkpoint= --device-id 3 --dir exp/chain/train_modelmodel_opadam_b
s128_ep6_lr5e-4_fpe150_110_90_hn625_fpr1500000_ms1_2_3_4_5 --feat-dim 43 --hidden-dim 625 --is-training true --kernel-size-list 1, 3
, 3, 3, 3, 3 --log-level info --output-dim 4336 --stride-list 1, 1, 3, 1, 1, 1 --multi-step 1, 2, 3, 4, 5 --model-name model --train
.cegs-dir exp/chain/merged_egs --train.den-fst exp/chain/den.fst --train.egs-left-context 13 --train.egs-right-context 13 --train.l2
-regularize 5e-4 --train.lr 5e-4 --train.num-epochs 6
2020-01-14 15:18:22,849 WARNING [train3.py:261] Done 
```

**Comment by csukuangfj** [comment]
@fanlu 
thanks

File: egs/aishell/s10/RESULTS [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# If you execute `run.sh`, then you should get similar results as follows:
+
+# Results for kaldi pybind LF-MMI training with PyTorch
+## head exp/chain/decode_res/*/scoring_kaldi/best_* > RESULTS
+#
+==> exp/chain/decode_res/dev/scoring_kaldi/best_cer <==
```
Comment (danpovey) [inline egs/aishell/s10/RESULTS]: The naming scheme is not obvious from this file... what is "res"?  Please clarify this, and also chain_nnet3.
And can you please make sure that these results (and where appropriate, the output of chain_dir_info.pl) are 
in a comment at the top of the script that generated them?

File: egs/aishell/s10/local/run_tdnn_1b.sh [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+  feat_dim=$(feat-to-dim scp:data/fbank_pitch/train/feats.scp -)
+
+  mkdir -p $dir/configs
+  cat <<EOF > $dir/configs/network.xconfig
+  input dim=$feat_dim name=input
+
+  # please note that it is important to have input layer with the name=input
+  # as the layer immediately preceding the fixed-affine-layer to enable
+  # the use of short notation for the descriptor
+  fixed-affine-layer name=lda input=Append(-1,0,1) affine-transform-file=$dir/configs/lda.mat
```
Comment (danpovey) [inline egs/aishell/s10/local/run_tdnn_1b.sh]: How much do we lose from removing i-vectors?  If you could make a comparison with run_tdnn_1a.sh via compare_wer.sh and put it in a comment at the top, that would be ideal.   (If there is no compare_wer.sh, please
see if someone over there can make one for this setup!).

File: egs/aishell/s10/local/run_tdnn_1b.sh [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+  steps/nnet3/xconfig_to_configs.py --xconfig-file $dir/configs/network.xconfig --config-dir $dir/configs/
+fi
+
+if [[ $stage -le 4 ]]; then
+  steps/nnet3/chain/train.py --stage $train_stage \
+    --cmd $cuda_cmd \
+    --feat.cmvn-opts "--norm-means=false --norm-vars=false" \
+    --chain.xent-regularize $xent_regularize \
+    --chain.leaky-hmm-coefficient 0.1 \
+    --chain.l2-regularize 0.00005 \
```
Comment (danpovey) [inline egs/aishell/s10/local/run_tdnn_1b.sh]: BTW, these days we tend to set chain.l2-regularize to zero and instead rely on l2 regularization in the TDNN or TDNN-F layers.  This reminds me that this recipe is super old!   Does someone at mobvoi have time to test out a more recent recipe?  E.g. you could try out the current Swbd recipe (I don't remember how much data is in aishell).  We need to make sure that we are comparing against a recent baseline, or we won't be aiming for the right place!!

File: egs/aishell/s10/RESULTS (hunk 1 above)
Comment (csukuangfj) [reply egs/aishell/s10/RESULTS]: thanks, I will change it to follow the current style of `egs/swbd/s5c/RESULTS`.

File: egs/aishell/s10/local/run_tdnn_1b.sh (hunk 2 above)
Comment (csukuangfj) [reply egs/aishell/s10/local/run_tdnn_1b.sh]: I did not use `ivector` since I have not figured it out how to integrate it into PyTorch.
I will try to add `ivector` and compare the results with/without using `ivector`.

File: egs/aishell/s10/local/run_tdnn_1b.sh (hunk 3 above)
Comment (csukuangfj) [reply egs/aishell/s10/local/run_tdnn_1b.sh]: No problem, I will switch to follow the recipe in swbd.

**Comment by danpovey** [comment]
Just noticed this branch has conflicts.

**Comment by csukuangfj** [comment]
I will resolve the conflicts tomorrow and try the new recipes during the Chinese New Year.

**Comment by danpovey** [comment]
Thanks!!!
BTW check with Haowen before trying the new recipes... I think he  was
going to try that.

**Comment by csukuangfj** [comment]
Conflicts resolved.

**Comment by qindazhu** [comment]
Have run latest recipe for aishell https://github.com/kaldi-asr/kaldi/pull/3868
